Logger

DannyPeck opened the pull request:
- Removed level parameter to log method
- Added a default log level property
- Created OutputVisitor
- StringVisitor is now defined as OutputVisitor<std::string>
- renamed logAtLevel to logIfAboveThreshold
- Moved Output related classes and interfaces in outputs/ directory and namespace

DannyPeck committed:
renamed internal log method; added default Level support

DannyPeck committed:
implemented OutputVisitor; removed private StringVisitor; created Outputs namespace/dir

DannyPeck committed:
removed object file

DannyPeck committed:
added comment about default level

DannyPeck committed:
reduced necessary variant includes within Outputable

DannyPeck committed:
reorganized tests directory; updated makefile to handle tests; added tests for Stringifiable

DannyPeck committed:
removed exe's

technetos commented on C++/Logger/Logger.cpp:
```diff
@@ -7,46 +7,47 @@ namespace Toolbox::Log
+void Logger::log (const String & message) const
 {
-  logAtLevel (level, message);
+  logIfAboveThreshold (_default, message);
```
`log_impl (message);`

log_impl internally should check if the level is above the threshold

technetos commented on C++/Logger/Logger.cpp:
```diff
@@ -59,7 +60,17 @@ void Logger::setThreshold (const Level & threshold)
+
+void Logger::logIfAboveThreshold (const Level & level, const String & message) const
 {
   if (isAboveThreshold (level))
```
```
if (isAboveThreshold (this->_default))
{
// do stuff
}
```

technetos commented on C++/Logger/Logger.cpp:
```diff
@@ -7,46 +7,47 @@ namespace Toolbox::Log
 void Logger::trace (const String & message) const
 {
-  logAtLevel (Level::TRACE, message);
+  logIfAboveThreshold (Level::TRACE, message);
```
`log (Level::TRACE, message);`

technetos commented on C++/Logger/Logger.cpp:
```diff
@@ -7,46 +7,47 @@ namespace Toolbox::Log
+void Logger::log (const String & message) const
 {
-  logAtLevel (level, message);
+  logIfAboveThreshold (_default, message);
```
`log` should be overloaded to take a message or a log level and a message:

`log (message);`
`log (_default, message);`

The `log` overload that does not take a level argument should call the log overload that does and should pass in `_default` as the log level

DannyPeck replied:
Good idea.

technetos commented on C++/Logger/Logger.cpp:
```diff
@@ -7,46 +7,47 @@ namespace Toolbox::Log
 void Logger::debug (const String & message) const
 {
-  logAtLevel (Level::DEBUG, message);
+  logIfAboveThreshold (Level::DEBUG, message);
```
`log (Level::DEBUG, message);`

DannyPeck replied:
I chose logIfAboveThreshold for its readability in the code.  Functions are meant to do one thing and you should know what it does without having to dive into the details.  Since logIfAboveThreshold is a private function it doesn't hurt or look weird to users of the interface.

technetos replied:
I half want to pass in a predicate function instead of hard coding the logic

DannyPeck replied:
What would that gain us?

technetos commented on C++/Logger/Logger.cpp:
```diff
@@ -59,7 +60,17 @@ void Logger::setThreshold (const Level & threshold)
+  _default = level;
+}
+
+void Logger::logIfAboveThreshold (const Level & level, const String & message) const
```
`void Logger::log (const Level & level, const String & message) const`

DannyPeck committed:
redesigned Output System to more closely represent its intent, conversions.  Outputable has become ConvertibleTo

DannyPeck committed:
commiting additional files to conversion commit

DannyPeck commented:
- Redesigned the Output/Outputable system to ConvertibleTo.  ConvertibleTo more closely captures the intent of what Outputable was attempting to do.  The idea is that you would define things in terms of whether they are "ConvertibleTo<Type>".  Example: class Stringifiable : public ConvertibleTo<Type>.  We are using ConvertibleTo to define different Conversion Interfaces.

- the output method has been renamed to "convert".
- added a ConversionVisitor which is a boost::variant static visitor designed to visit variants of type <ResultType, ConvertibleTo<ResultType>>.  Ideally you could define a ConversionVariant that embodies the variant of <ResultType, ConvertibleTo<ResultType>> and encapsulates away the use of the Visitor into either an implicit/explicit conversion or a get () method.  Though this was causing problems.

DannyPeck committed:
fixed tests; fixed issues with codacy

DannyPeck committed:
Added Codacy Badge

DannyPeck committed:
Merge pull request #17 from CachedNerds/issue13-codacy

Issue13 codacy

DannyPeck committed:
Created ConversionVariant type trait to encapsulate the boost variant declaration; this is necessary as the constness of the types needs to be handles by the Conversion interfaces

DannyPeck committed:
Merge branch 'logger' of https://github.com/CachedNerds/Toolbox into logger

DannyPeck committed:
Created ConversionVisitor tests

DannyPeck committed:
Renamed ConversionVariant to Variant & ConversionVisitor to VariantVisitor; added static get method to the Variant class for applying the VariantVisitor

DannyPeck committed:
updated tests to reflect Conversions Variant change

DannyPeck committed:
renamed directories to be singular; removed remaining references to the visitors within the logger and basic sink

DannyPeck commented:
- Simplified naming, don't need Conversion prefix if it is within the Conversion namespace
- Added Variant type trait that defines the boost::variant correctly and abstracts it's use from the user.
- Added static get method to Variant for applying the visitor, makes using the Variant more user friendly.
